lib/compilers/java.js

```diff
@@ -124,8 +127,36 @@ export class JavaCompiler extends BaseCompiler {
         return result;
     }
 
-    getMainClassName() {
-        // TODO(supergrecko): The main class name is currently hardcoded
+    async getMainClassName(dirPath) {
+        const maxSize = this.env.ceProps('max-asm-size', 64 * 1024 * 1024);
+        const files = await fs.readdir(dirPath);
+        const results = await Promise.all(
+            files
+                .filter(f => f.endsWith('.class'))
+                .map(async classFile => {
+                    const options = {
+                        maxOutput: maxSize,
+                        customCwd: dirPath,
+                    };
+                    const objResult = await this.exec(this.compiler.objdumper, [classFile], options);
+                    if (objResult.code !== 0) {
+                        return null;
+                    }
+
+                    if (objResult.stdout.includes('void main(java.lang.String[])')) {
```

Comment:
Can we make this `public static void main(java.lang.String[])`?

Reply:
It's possible for the main method to be marked `final`, this check would have to be expanded to `public static void main(java.lang.String[])` with an additional check for `public static final void main(java.lang.String[])`, if that's okay.

Reply:
In that case we should also account for strictfp and synchronized keywords, making the final check list 
```java
public static final synchronized strictfp void main(java.lang.String[])
public static final synchronized void main(java.lang.String[])
public static final void main(java.lang.String[])
public static void main(java.lang.String[])
```

Reply:
At this point would bringing a regex into the mix be appropriate? 

Something like `public static ?(?s)(.*?) void main\(java\.lang\.String\[\]\)`

Reply:
Wouldn't that be

public static final synchronized strictfp void main(java.lang.String[])
public static final synchronized void main(java.lang.String[])
public static final strictfp void main(java.lang.String[])
public static final void main(java.lang.String[])
public static synchronized strictfp void main(java.lang.String[])
public static synchronized void main(java.lang.String[])
public static strictfp void main(java.lang.String[])
public static void main(java.lang.String[])

?

To avoid looking for O(2^n) strings, I vote something like objResult.stdout.replaceAll(" final ", " ").replaceAll(" synchronized ", " ").replaceAll(" strictfp ", " ").contains("public static void main(java.lang.String[])")

Reply:
I believe javap orders the modifiers so that wouldn't be necessary, but RegEx would also be a sufficient solution